File: src/main/python/leetcode_by_python/SegmentOne/No33_Search_in_Rotated_Sorted_Array.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/28 20:10
# https://leetcode.com/problems/search-in-rotated-sorted-array/
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid] == target:
                return mid
            if nums[l] <= nums[mid]:
                if nums[l] <= target < nums[mid]:
                    r = mid - 1
                else:
                    l = mid + 1
            else:
                if nums[mid] < target <= nums[r]:
                    l = mid + 1
                else:
                    r = mid - 1

        return -1


if __name__ == '__main__':
    solution = Solution()
    assert solution.search([4, 5, 6, 7, 0, 1, 2], 0) == 4
    assert solution.search([4, 5, 6, 7, 0, 1, 2], 3) == -1

File: src/main/java/com/hazza/leetcode/SegmentOne/No33SearchInRotatedSortedArray.java
package com.hazza.leetcode.SegmentOne;

/**
 * Created by hazza on 7/18/17.
 */
public class No33SearchInRotatedSortedArray {
    public int search(int[] nums, int target) {
        if (nums == null || nums.length == 0) return -1;
        int left = 0, right = nums.length - 1;
        while (left < right) {
            int middle = (left + right) / 2;
            if (nums[middle] == target) return middle;
            if (nums[left] <= nums[middle]) {
                if (target >= nums[left] && target < nums[middle]) right = middle - 1;
                else left = middle + 1;
            } else {
                if (target > nums[middle] && target <= nums[right]) left = middle + 1;
                else right = middle - 1;
            }
        }

        return nums[left] == target ? left : -1;
    }
}
